### PR TITLE
HDDS-5839. Make sure buckets created from OFS are in FILE_SYSTEM_OPTI…

### PR DESCRIPTION
**Pull request by JyotinderSingh**
…MIZED layout

## What changes were proposed in this pull request?

The idea of this task is to change the default layout behavior of ofs client, it should by default create bucket in FILE_SYSTEM_OPTIMIZED layout.

Presently, client is sending BucketLayout.DEFAULT, which is LEGACY to the OM. This has to be modified to FILE_SYSTEM_OPTIMIZED.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5839

## How was this patch tested?

Related integration tests

### CODE REVIEW

**Comment by JyotinderSingh** [comment]
I've updated the patch with the requested changes @rakeshadr @bharatviswa504 
Currently on my local I can see partial test failures in `TestOzoneFileInterfaces.java`, `TestOzoneFSWithObjectStoreCreate.java,` `TestRootedOzoneFileSystem.java`, `TestOzoneFileSystem.java`. These seem to mainly be due to some behaviour changes introduced by changes.

**Comment by JyotinderSingh** [comment]
Mentioning here for reference: I am adding the client side validation back to the patch which will throw an exception if OFS encounters a bucket with OBS layout. This requires some workarounds in some of the tests where we need LEGACY bucket layouts (we disallow that operation directly - [layout defaults to OBS if client tries to create a LEGACY bucket](https://github.com/apache/ozone/blob/4c2181ff42be3d6b7dbec4f0c3da549d89e12fc0/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/OMBucketCreateRequest.java#L145)).

To obtain a LEGACY bucket we set the following config in the MiniOzoneCluster:
```
conf.set(OMConfigKeys.OZONE_DEFAULT_BUCKET_LAYOUT, BucketLayout.LEGACY.name());
```

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 1]
```diff
@@ -2993,6 +2993,21 @@
...
+      future.
+    </description>
+  </property>
+
+  <property>
+    <name>ozone.client.test.ofs.default.bucket.layout</name>
+    <value>FILE_SYSTEM_OPTIMIZED</value>
+    <tag>OZONE, MANAGEMENT</tag>
+    <description>
+      Default bucket layout used by Ozone Manager when buckets are created using Ozone File System.
```
Comment (rakeshadr) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:3006]: Please change the description to convey that this is client side config and for testing purpose only.
How about something like,

"`This configuration is used for testing purpose. Default bucket layout value when buckets are created using ofs. Supported values are LEGACY and FILE_SYSTEM_OPTIMIZED`"

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 2]
```diff
@@ -2993,6 +2993,21 @@
...
       will create intermediate directories.
+      For testing purposes LEGACY layout is also allowed. LEGACY layout is not recommended and will be removed in the
+      future.
+    </description>
+  </property>
+
+  <property>
+    <name>ozone.client.test.ofs.default.bucket.layout</name>
+    <value>FILE_SYSTEM_OPTIMIZED</value>
+    <tag>OZONE, MANAGEMENT</tag>
```
Comment (rakeshadr) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:3004]: tag should be :

`<tag>OZONE, CLIENT</tag>`

**Comment by rakeshadr** [comment]
> Instead of changing the default value from OBS to LEGACY, I am thinking of introducing a client side flag
> 
> **option-1)** `ozone.client.bucket.layout.validation=false`. By default we can set it to false and once the cluster is fully upgraded customer can set it to true that will enable the o3fs/ofs validation check.
> 
> **option-2)** `ozone.client.bucket.layouts.allowed=FSO,OBS`. Here a comma-separated list of values where ofs/o3fs can do a validation check. Later after upgradation this can be set to FSO.

The client side validation will be handled in jira HDDS-5959 and will continue the discussion there.

**Comment by rakeshadr** [comment]
+1 LGTM.

 Thank you @JyotinderSingh for the good work and will merge it once we get a clean CI report.

Thanks @mukul1987 , @bharatviswa504 , @aryangupta1998 for the reviews.

**Pull request merged**